convex-core/src/main/java/convex/core/lang/reader/AntlrReader.java

```diff
@@ -471,12 +482,101 @@ public static AList<ACell> readAll(CharStream cs) {
 		ConvexParser parser = new ConvexParser(tokens);
 		parser.removeErrorListeners();
 		ParseTree tree = parser.forms();
-		
+
 		CRListener visitor=new CRListener();
 		ParseTreeWalker.DEFAULT.walk(visitor, tree);
-		
+
 		ArrayList<ACell> top=visitor.popList();
 		return Lists.create(top);
 	}
 
+	public static ACell readOne(java.io.PushbackReader r) {
+		try {
+			CharStream cs = new InteractiveCharStream(r);
+			ConvexLexer lexer=new ConvexLexer(cs);
+			lexer.removeErrorListeners();
+			lexer.setTokenFactory(new CommonTokenFactory(true));
+			TokenStream tokens = new InteractiveTokenStream(lexer);
+			ConvexParser parser = new ConvexParser(tokens);
+			parser.removeErrorListeners();
+
+			ParseTree tree = parser.form();
+
+			r.unread(cs.LA(1));
+
+			CRListener visitor=new CRListener();
+			ParseTreeWalker.DEFAULT.walk(visitor, tree);
+
+			ArrayList<ACell> top=visitor.popList();
+			if (top.size()!=1) {
+				throw new ParseException("Bad parse output: "+top);
+			}
+
+			return top.get(0);
+
+		} catch (IOException e) {
+			throw Utils.sneakyThrow(e);
+		}
+
+	}
+
+
+
+}
+
+class InteractiveCharStream extends UnbufferedCharStream {
+
+	InteractiveCharStream(java.io.Reader input)
+	{
+		super(input);
+	}
+
+	@Override
+	public void consume() {
+
+		if (LA(1) == IntStream.EOF) {
+			throw new IllegalStateException("cannot consume EOF");
+		}
+
+		// buf always has at least data[p==0] in this method due to ctor
+		lastChar = data[p];   // track last char for LA(-1)
+		if (p == n-1 && numMarkers==0) {
+			n = 0;
+			p = -1; // p++ will leave this at 0
+			lastCharBufferStart = lastChar;
+		}
+
+		p++;
+		currentCharIndex++;
+		//sync(1);
```

Comment:
These two consume implementations are the same as the class the `UnbufferedXStream` class they extend except without the `sync` call at the end, which consumes more characters/tokens from the input